internal/k8s/reconciler/state/gateway.go

```diff
@@ -0,0 +1,105 @@
+package state
+
+import (
+	"errors"
+
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/hashicorp/consul-api-gateway/internal/core"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/reconciler/common"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/reconciler/status"
+)
+
+// GatewayState holds ephemeral state for gateways
+type GatewayState struct {
+	ConsulNamespace string
+	Status          status.GatewayStatus
+	PodReady        bool
+	ServiceReady    bool
+	Generation      int64
+	Addresses       []string
+
+	Listeners []*ListenerState
+}
+
+func InitialGatewayState(g *gwv1beta1.Gateway) *GatewayState {
+	state := &GatewayState{
+		Generation: g.GetGeneration(),
+	}
+	for _, listener := range g.Spec.Listeners {
+		state.Listeners = append(state.Listeners, &ListenerState{
+			Name:     listener.Name,
+			Protocol: listener.Protocol,
+			Routes:   make(map[string]core.ResolvedRoute),
+		})
+	}
+	return state
+}
+
+func (g *GatewayState) GetStatus(gateway *gwv1beta1.Gateway) gwv1beta1.GatewayStatus {
+	listenerStatuses := []gwv1beta1.ListenerStatus{}
+	for _, state := range g.Listeners {
+		listenerStatuses = append(listenerStatuses, state.getStatus(g.Generation))
+	}
+
+	conditions := g.Status.Conditions(g.Generation)
+
+	// prefer to not update to not mess up timestamps
+	if status.ListenerStatusesEqual(listenerStatuses, gateway.Status.Listeners) {
+		listenerStatuses = gateway.Status.Listeners
+	}
+	if status.ConditionsEqual(conditions, gateway.Status.Conditions) {
+		conditions = gateway.Status.Conditions
+	}
+
+	ipType := gwv1beta1.IPAddressType
+	addresses := make([]gwv1beta1.GatewayAddress, 0, len(g.Addresses))
+	for _, address := range g.Addresses {
+		addresses = append(addresses, gwv1beta1.GatewayAddress{
+			Type:  &ipType,
+			Value: address,
+		})
+	}
+
+	return gwv1beta1.GatewayStatus{
+		Addresses:  addresses,
+		Conditions: conditions,
+		Listeners:  listenerStatuses,
+	}
+}
+
+// ListenerState holds ephemeral state for listeners
+type ListenerState struct {
```

Comment:
I don't think that this is currently in-use? IIRC the original PR did some modification of the K8sListener struct to try and tie the `ListenerState` onto the `K8sListener` before removing the `K8sListener` in the follow-ups. Without this in use I think there might be issues trying to call things like the `GatewayState.GetStatus` which calculates statuses based off of this state.

Reply:
Good catch!

This struct currently gets consumed up in `GatewayState` and now, with commits since you reviewed, is initialized on the `K8sListener` as part of `NewK8sGateway`. I believe this leaves the only difference from the [original refactor](https://github.com/hashicorp/consul-api-gateway/pull/160/files#diff-53c61379ab3051a987420aef6574a34645539c538cd1250d49a93b584667724b) being that this logic moves into the `Factory`. That change is incoming on the PR after this one.

Lemme know if ya think I've missed something

Reply:
So, just be aware that by the time the final PR rolled around, the factory actually got completely dropped since it was no longer necessary. If you have any ideas to avoid the intermediate introduction of the factories while keeping this fairly clean/in test-passable chunks, feel free to drop their introduction. Otherwise, I'm fine with going the same route as before just so we don't have to think too hard about this 😅